### PR TITLE
313 post banner attachment

### LINKED ISSUE #313: [313] Post Banner Attachment
**Current state**
There is no way to add a post banner and/or crop it.

**Expected state**
The author of a post should be able to crop their post banner and also change it by uploading a file to our server. The foundation of this functionality has been made in #220.

**TO DO**
- [ ] Upload image (Drag 'n drop & clickable)
- [ ] Crop image

**Necessities**
#220

**Branch**
313-profile-banner-attachment

### PR DESCRIPTION
**Pull request by jerohero**
closes #313 

<br />

 **What is it supposed to do**  
This branch adds functionality for post thumbnails/banners. This is the image the user sees with the post on places like the homepage, and on the top of the page when the user visits the page of a post. The owner of a post can now change the banner by uploading an image to the server. The user is able to do this through the application by selecting or dropping an image. This image can be cropped.

image

<br />

 **How can we test your branch**  
Go to one of your posts or create a new one. Hovering over the background thumbnail should show _edit banner_, which can be clicked. This should not happen when you're not logged in as the owner of the post. Only .png or .jpg files should be allowed. Max file size should be 4000 KB. You should only be able to save an image after cropping it. The new image should appear immediately as the new thumbnail, and this should still be there after refreshing. The new image should also be on the homepage (as a thumbnail for the post). Try changing the thumbnail again. Updating the image should delete the old one (from the file system and Attachment table).

<br />

 **Additional notes**  
**You'll have to add some folders to make this branch work**. In _backend/app/public_, add _post/thumbnail_.

Wasn't sure whether I should completely delete the thumbnail after a post is archived or not, so at the moment the thumbnail attachments don't get removed when the post is archived.

When fetching multiple posts, a separate DAO call is executed after for every single post in order to fetch their banner attachments. Maybe there's a more efficient way to do this?
image

<br />

**1. General checklist**
- [x] a. Tested happy flow.
- [x] b. Tested unhappy flow.
- [x] c. No unexpected exceptions.
- [x] d. No code with high complexity. (Big O)
- [x] e. EOF newline.
- [x] f. No linter warnings or errors.

<br />

**2. If relevant, front-end checklist**
- [x] a. Responsive. (Desktop all the way to mobile 320px)
- [x] b. Using SCSS variables where relevant.
- [x] c. Relevant class names
- [x] d. No modified global SCSS properties, except if necessary and fully tested the impact.
- [x] e. Checked [caniuse](https://caniuse.com) for more modern CSS properties.

<br />

**3. If relevant, back-end checklist**
- [x] a. No unhandled promise rejections.
- [x] b. Seeders still functional.
- [x] c. No primary-key violations.
- [x] d. Normalization, at least [3NF](https://en.wikipedia.org/wiki/Database_normalization).
- [x] e. Used guards (middleware) where necessary and tested if route is guarded.

<br />

**4. If relevant, test these browsers**
- [x] a. Chrome
- [x] b. Firefox
- [ ] c. Opera
- [x] d. Edge
- [ ] e. Safari

### CODE REVIEW

**Review by AjobK — CHANGES_REQUESTED** [review]
## Issue 1
When there is no /post/thumbnail folder you can't add images to that folder. You have to make a .gitkeep file so folders like this always stay. I've made it for you this time, so you can just `git pull`.

## Issue 2
When I'm trying to save a thumbnail on a post without a title or content, it gives me the following error:

image

## Issue 3
Always create a placeholder banner instead of this grey blob. When you save a post without a banner it gets a default banner, which is good. Why not show this banner on post creation too?

image

File: frontend/src/components/postLike/index.js [hunk 1]
```diff
@@ -77,7 +76,7 @@ class PostLike extends Component {
         {likesAmount <= 0 && <p className={ `${styles.postLikesAmount}` }>0 likes</p>}
         {likesAmount > 0 && (
           <p className={ `${styles.postLikesAmount} ${styles.clickableLikes}` } onClick={ this.openLikesList }>
-            {unitFormatterUtil.getNumberSuffix(likesAmount)} {likesAmount === 1 ? 'like' : 'likes'}
+            {UnitFormatterUtil.getNumberSuffix(likesAmount)} {likesAmount === 1 ? 'like' : 'likes'}
```
Comment (AjobK) [inline frontend/src/components/postLike/index.js]:
```suggestion
            { UnitFormatterUtil.getNumberSuffix(likesAmount)} {likesAmount === 1 ? 'like' : 'likes' }
```

File: frontend/src/components/postBanner/index.js [hunk 2]
```diff
@@ -57,6 +142,16 @@ class PostBanner extends Component {
...
+          { this.state.upThumbnail && (
+            <Cropper
+              inputType={ 'thumbnail' }
+              img={ this.state.upThumbnail }
+              entityId={ this.props.post.path }
+              closeCropper={ this.closeCropper }
+              changeImage={ this.changeThumbnail }
+              returnOnSave={ this.props.isNew }
+            />
+          )}
```
Comment (AjobK) [inline frontend/src/components/postBanner/index.js]:
```suggestion
          ) }
```

File: frontend/src/components/postBanner/index.js [hunk 3]
```diff
@@ -1,14 +1,95 @@
...
+      this.setState({
+        thumbnail: newThumbnail
+      })
+
+      return
+    }
+
+    this.props.onThumbnailAdded(newThumbnail)
+
+    const reader = new FileReader()
```
Comment (AjobK) [inline frontend/src/components/postBanner/index.js:66]:
```suggestion
    const reader = new FileReader()
    
```

File: frontend/src/components/postViews/index.js [hunk 4]
```diff
@@ -34,7 +33,7 @@ class PostViews extends Component {
       <div className={ styles.postViews }>
         <Icon iconName={ 'Eye' } className={ styles.viewIcon } />
         <p className={ styles.postViewsText }>
-          {`${unitFormatterUtil.getNumberSuffix(this.state.views)} 
+          {`${UnitFormatterUtil.getNumberSuffix(this.state.views)} 
                     ${this.state.views === 1 ? 'view' : 'views'}`}
```
Comment (AjobK) [inline frontend/src/components/postViews/index.js]: Strangely formatted code? Maybe use something like this:

```suggestion
          { `
              ${ UnitFormatterUtil.getNumberSuffix(this.state.views) } 
              ${ this.state.views === 1 ? 'view' : 'views' }
          `}
```

**Review by AjobK — CHANGES_REQUESTED** [review]
Pure code comments. Gonna test now.

**Comment by AjobK** [comment]
image

Error when creating empty post and then editing it

**Comment by AjobK** [comment]
image

Able to change post banner of a post that I'm not an author of (through insomnia/postman).

File: backend/app/controllers/postController.ts [hunk 5]
```diff
@@ -6,24 +6,32 @@ import AccountDAO from '../daos/accountDAO'
 import PostLike from '../entities/post_like'
 import Profile from '../entities/profile'
 import ProfileDAO from '../daos/profileDAO'
+import AttachmentDAO from '../daos/attachmentDAO'
+import attachmentDAO from '../daos/attachmentDAO'
```
Comment (ryankroon00) [inline backend/app/controllers/postController.ts]: 2 keer attachment dao, 1tje verwijderen

File: frontend/src/components/cropper/index.js [hunk 6]
```diff
@@ -132,6 +155,35 @@ class Cropper extends Component {
       })
   }
 
+  sendImageToDb = async (image) => {
```
Comment (ryankroon00) [inline frontend/src/components/cropper/index.js]: You don't send it directly to the db so I would call this function something different

File: frontend/src/components/userBanner/userBanner.scss [hunk 7]
```diff
@@ -63,6 +63,7 @@
 		input {
 			cursor: pointer;
 			width: 100%;
```
Comment (ryankroon00) [inline frontend/src/components/userBanner/userBanner.scss]: width: 100% twice

File: frontend/src/pages/post/index.js [hunk 8]
```diff
@@ -122,23 +122,44 @@ class Post extends App {
   sendToDB = (path = null) => {
     Axios.defaults.baseURL = this.props.store.defaultData.backendUrl
 
+    if (!path) {
+      this.createPostDB()
+    } else if (typeof path == 'string') {
+      this.updatePostDB()
```
Comment (ryankroon00) [inline frontend/src/pages/post/index.js]: I would remove db from the name

**Review by ryankroon00 — CHANGES_REQUESTED** [review]

**Comment by AjobK** [comment]
Has merge conflicts

**Comment by jerohero** [comment]
> image
> 
> Able to change post banner of a post that I'm not an author of (through insomnia/postman).

@AjobK I assume you were logged in as an administrator with post permissions. Should I disable administrators having access to posts?

**Comment by AjobK** [comment]
For now, yes. We will make a seperate issue in which this will be implemented in a more specific approach.

Admins shouldnt be able to change a posts banner, but remove it. Changing it would alter a users post intentions.

**Review by AjobK — CHANGES_REQUESTED** [review]
Check comment above

**Review by AjobK — APPROVED** [review]

**Review by ryankroon00 — APPROVED** [review]

**Pull request merged**